content/articles/ansible-and-nix.md

```diff
@@ -0,0 +1,333 @@
+
+NixOS is a simple and modern Linux distribution configured declaratively using
+the same language as Nix packages.
+
+Finally, NixOps is used to configure a «network» of NixOS systems, still using
+the same language.
+
+The learning curve for those tools is steep, the official documentation is
+light and there aren't that many resources available on the web about Nix so I
+had to learn it mostly by reading the source of the tools. Also, NixOps being a
+young project I hit a few open bugs.
+
+The [Nix manual](https://nixos.org/nix/manual/) is a good start to learn about
+the package manager. This [Nix by example
+article](https://medium.com/@MrJamesFisher/nix-by-example-a0063a1a4c55)
+describes the Nix expression language in more detail.
+To understand the rationale behind Nix and NixOS, you can read the accessible
+[NixOS: A Purely Functional Linux
+Distribution](http://nixos.org/~eelco/pubs/nixos-jfp-final.pdf) paper.
+
+For more practical questions, I found that the best way to learn was to find a
+package that resembles what I wanted to do and get inspiration from its
+corresponding expression. NixOS packages can be found using the [Search NixOS
+packages page](https://nixos.org/nixos/packages.html).
+
+
+## Comparing the deployment of a test application
+
+Using the [test
+project](https://github.com/WeAreWizards/voting-deployment-test), I am going to
+illustrate the practical differences between using NixOps and Ansible.
+
+### Dependencies
+
+Nix is first and foremost a package manager, so the most common way to deploy
+applications is to build packages locally and copy them to servers.
+
+Nix being `purely functional`, it should build the same output (a closure)
+given the same input (an expression). To do so, Nix needs all dependencies
+retrieved at build time.
+
+Assuming the dependencies you need are present in the nixos channel and at the
+right version then you can just refer to them like in the main derivation of
+the test project:
+
+```nix
+pythonPackages.buildPythonPackage {
+    name = "voting";
+
+    propagatedBuildInputs = [
+        pkgs.python27Packages.django
+        pkgs.python27Packages.sqlite3
+        pkgs.python27Packages.gunicorn
+    ];
+
+    src = ./.;
+}
+```
+
+However, if your dependencies are not in the nixos channel, you will need to
+declare them manually.
+
+For example the requests python library is declared like this:
+
+```nix
+requests = buildPythonPackage rec {
+  name = "requests-1.2.3";
+
+  src = pkgs.fetchurl {
+    url = "http://pypi.python.org/packages/source/r/requests/${name}.tar.gz";
+    md5 = "adbd3f18445f7fe5e77f65c502e264fb";
+  };
+
+  meta = {
+    description = "An Apache2 licensed HTTP library, written in Python, for human beings";
+    homepage = http://docs.python-requests.org/en/latest/;
+  };
+};
+```
+
+Which means than when building the package you are still dependent on pypi
+being up, unless the package is already cached.
```

Comment:
Channels are just the metadata, packages are still downloaded from e.g. pipy. But because you know the hash of the source you can just mirror it arbitrarily, e.g. onto your own infrastructure and then use a mirror:// URL like this: https://github.com/NixOS/nixpkgs/blob/cc4a5073e2c26edbf5f74c03c8fb169b5ff0a8df/pkgs/development/libraries/openjpeg/2.0.1.nix#L7